content/graph/WeightedMatching.h

```diff
@@ -1,89 +1,44 @@
+ * License: CC0
+ * Source: https://github.com/bqi343/USACO/blob/master/Implementations/content/graphs%20(12)/Matching/Hungarian.h
+ * Description: Minimum cost bipartite matching. Takes in cost[N][M], where N
+ * <= M, and cost[i][j] = cost for i-th node on left to j-th node on right.
+ * Returns flow and match, where match[i] = node assigned to i-th left node.
+ * Negate costs for max cost.
+ * Time: O(N^2M)
+ * Status: Tested on kattis:cordonbleu
  */
 #pragma once
 
-typedef vector<double> vd;
-bool zero(double x) { return fabs(x) < 1e-10; }
-double minCostMatching(const vector<vd>& cost, vi& L, vi& R) {
-	int n = sz(cost), mated = 0;
-	vd dist(n), u(n), v(n);
-	vi dad(n), seen(n);
-
-	/// construct dual feasible solution
-	rep(i,0,n) {
-		u[i] = cost[i][0];
-		rep(j,1,n) u[i] = min(u[i], cost[i][j]);
-	}
-	rep(j,0,n) {
-		v[j] = cost[0][j] - u[0];
-		rep(i,1,n) v[j] = min(v[j], cost[i][j] - u[i]);
-	}
-
-	/// find primal solution satisfying complementary slackness
-	L = R = vi(n, -1);
-	rep(i,0,n) rep(j,0,n) {
-		if (R[j] != -1) continue;
-		if (zero(cost[i][j] - u[i] - v[j])) {
-			L[i] = j;
-			R[j] = i;
-			mated++;
-			break;
-		}
-	}
-
-	for (; mated < n; mated++) { // until solution is feasible
-		int s = 0;
-		while (L[s] != -1) s++;
-		fill(all(dad), -1);
-		fill(all(seen), 0);
-		rep(k,0,n)
-			dist[k] = cost[s][k] - u[s] - v[k];
-
-		int j = 0;
-		for (;;) { /// find closest
-			j = -1;
-			rep(k,0,n){
-				if (seen[k]) continue;
-				if (j == -1 || dist[k] < dist[j]) j = k;
+pair<int, vi> hungarian(const vector<vi> &a) {
+	int n = sz(a) + 1, m = sz(a[0]) + 1;
+	vi u(n), v(m), p(m), ans(n - 1);
+	rep(i,1,n) {
+		p[0] = i;
+		int j0 = 0; // add "dummy" worker 0
+		vi dist(m, INT_MAX), pre(m, -1);
+		vector<bool> done(m + 1);
+		do { // dijkstra
+			done[j0] = true;
+			int i0 = p[j0], j1, delta = INT_MAX;
+			rep(j,1,m) if (!done[j]) {
+				auto cur = a[i0 - 1][j - 1] - u[i0] - v[j];
+				if (cur < dist[j]) dist[j] = cur, pre[j] = j0;
+				if (dist[j] < delta) delta = dist[j], j1 = j;
 			}
-			seen[j] = 1;
-			int i = R[j];
-			if (i == -1) break;
-			rep(k,0,n) { /// relax neighbors
-				if (seen[k]) continue;
-				auto new_dist = dist[j] + cost[i][k] - u[i] - v[k];
-				if (dist[k] > new_dist) {
-					dist[k] = new_dist;
-					dad[k] = j;
-				}
+			rep(j,0,m) {
+				if (done[j]) u[p[j]] += delta, v[j] -= delta;
+				else dist[j] -= delta;
 			}
+			j0 = j1;
+		} while (p[j0]);
+		while (j0) { // update alternating path
+			int j1 = pre[j0];
+			p[j0] = p[j1], j0 = j1;
 		}
-
-		/// update dual variables
-		rep(k,0,n) {
-			if (k == j || !seen[k]) continue;
-			auto w = dist[k] - dist[j];
-			v[k] += w, u[R[k]] -= w;
-		}
-		u[s] += dist[j];
-
-		/// augment along path
-		while (dad[j] >= 0) {
-			int d = dad[j];
-			R[j] = R[d];
-			L[R[j]] = j;
-			j = d;
-		}
-		R[j] = s;
-		L[s] = j;
 	}
-	auto value = vd(1)[0];
-	rep(i,0,n) value += cost[i][L[i]];
-	return value;
+	rep(j,1,m) if (p[j]) ans[p[j] - 1] = j - 1;
```

Comment:
maybe using `n`/`m` as the dummy nodes could make this shorter

Reply:
I think it's a bit uglier doing this.
```cpp
pair<int, vi> hungarian(const vector<vi> &a) {
    if (a.empty()) return {0, {}};
    int n = sz(a), m = sz(a[0]) + 1;
    vi u(n), v(m), p(m, -1), ans(n - 1);
    rep(i, 0, n) {
        p[m - 1] = i;
        int j0 = m - 1; // add "dummy" worker 0
        vi dist(m, INT_MAX), pre(m, -1);
        vector<bool> done(m + 1);
        do { // dijkstra
            done[j0] = true;
            int i0 = p[j0], j1, delta = INT_MAX;
            rep(j, 0, m - 1) if (!done[j]) {
                auto cur = a[i0][j] - u[i0] - v[j];
                if (cur < dist[j]) dist[j] = cur, pre[j] = j0;
                if (dist[j] < delta) delta = dist[j], j1 = j;
            }
            rep(j, 0, m) {
                if (done[j]) u[p[j]] += delta, v[j] -= delta;
                else dist[j] -= delta;
            }
            j0 = j1;
        } while (p[j0] != -1);
        while (j0 != m - 1) { // update alternating path
            int j1 = pre[j0];
            p[j0] = p[j1], j0 = j1;
        }
    }
    rep(j, 0, m - 1) if (p[j]) ans[p[j]] = j;
    return {-v[m - 1], ans}; // min cost
}
```